### PR TITLE
GH-40394: [C++] Add support for mold

### LINKED ISSUE #40394: [C++] Add support for mold
### Describe the enhancement requested

https://github.com/rui314/mold

It's a faster linker. It will reduce build time.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

mold is a faster linker. It will reduce build time.

mold supports only ELF. So it's available on only Linux and *BSD. 

### What changes are included in this PR?

Add new `ARROW_USE_MOLD` CMake option. It's `OFF` by default because it doesn't work with conda.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.
* GitHub Issue: #40394

### CODE REVIEW

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp -g linux

**Comment by pitrou** [comment]
I get the following warnings when building:
```
[407/859] Linking CXX shared library debug/libarrow.so.1600.0.0
mold: warning: /home/antoine/arrow/dev/cpp/src/arrow/symbols.map: cannot assign version `global` to symbol `descriptor_table_Flight_2eproto`: symbol not found
mold: warning: /home/antoine/arrow/dev/cpp/src/arrow/symbols.map: cannot assign version `global` to symbol `descriptor_table_FlightSql_2eproto`: symbol not found
[591/859] Linking CXX shared library debug/libarrow_cuda.so.1600.0.0
mold: warning: /home/antoine/arrow/dev/cpp/src/arrow/symbols.map: cannot assign version `global` to symbol `descriptor_table_Flight_2eproto`: symbol not found
mold: warning: /home/antoine/arrow/dev/cpp/src/arrow/symbols.map: cannot assign version `global` to symbol `descriptor_table_FlightSql_2eproto`: symbol not found
```

and then I get various kinds of crashes when running the tests.

**Comment by pitrou** [comment]
`/build/build-test/debug/arrow-acero-fetch-node-test` throws an exception here:
```backtrace
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737222772352) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140737222772352) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140737222772352, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ffff4314476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff42fa7f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ffff45f3f00 in __gnu_cxx::__verbose_terminate_handler () at ../../../../libstdc++-v3/libsupc++/vterminate.cc:95
#6  0x00007ffff45f243c in __cxxabiv1::__terminate (handler=<optimized out>) at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:48
#7  0x00007ffff45f248e in std::terminate () at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:58
#8  0x00007ffff45f2680 in __cxxabiv1::__cxa_throw (obj=0x55555561cd70, tinfo=0x7ffff47108d8 <typeinfo for std::system_error>, 
    dest=0x7ffff460fa28 <std::system_error::~system_error()>) at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:98
#9  0x00007ffff45ec92b in std::__throw_system_error (__i=-1)
    at /home/conda/feedstock_root/build_artifacts/gcc_compilers_1706816862910/work/build/x86_64-conda-linux-gnu/libstdc++-v3/include/system_error:233
#10 0x000055555559dcc3 in std::call_once<arrow::acero::RegisterTestNodes()::<lambda()> >(std::once_flag &, struct {...} &&) (__once=..., __f=...)
    at /home/antoine/mambaforge/envs/pyarrow/x86_64-conda-linux-gnu/include/c++/12.3.0/mutex:860
#11 0x000055555559ac67 in arrow::acero::RegisterTestNodes () at /home/antoine/arrow/dev/cpp/src/arrow/acero/test_nodes.cc:360
#12 0x00005555555cc53a in arrow::acero::CheckFetch (options=...) at /home/antoine/arrow/dev/cpp/src/arrow/acero/fetch_node_test.cc:43
#13 0x00005555555cd976 in arrow::acero::FetchNode_Basic_Test::TestBody (this=0x55555561d320) at /home/antoine/arrow/dev/cpp/src/arrow/acero/fetch_node_test.cc:75
#14 0x00007ffff7d5f6d6 in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=0x55555561d320, method=&virtual testing::Test::TestBody(), 
    location=0x7ffff7d26f83 "the test body") at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:2665
#15 0x00007ffff7d39b6f in testing::Test::Run (this=0x55555561d320) at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:2682
#16 0x00007ffff7d3a625 in testing::TestInfo::Run (this=0x55555561eba0) at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:2861
#17 0x00007ffff7d3b011 in testing::TestSuite::Run (this=0x55555561d4b0) at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:3015
#18 0x00007ffff7d4b7d9 in testing::internal::UnitTestImpl::RunAllTests (this=0x55555561ca50) at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:5855
#19 0x00007ffff7d608c4 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x55555561ca50, 
    method=(bool (testing::internal::UnitTestImpl::*)(class testing::internal::UnitTestImpl * const)) 0x7ffff7d4b3c2 <testing::internal::UnitTestImpl::RunAllTests()>, 
    location=0x7ffff7d279e8 "auxiliary test code (environments or event listeners)") at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:2665
#20 0x00007ffff7d49be3 in testing::UnitTest::Run (this=0x7ffff7d80ec0 <testing::UnitTest::GetInstance()::instance>)
    at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:5438
#21 0x00007ffff7d82a02 in RUN_ALL_TESTS () at /build/build-test/_deps/googletest-src/googletest/include/gtest/gtest.h:2490
#22 0x00007ffff7d82a4b in main (argc=1, argv=0x7fffffffc0f8) at /build/build-test/_deps/googletest-src/googletest/src/gtest_main.cc:52
```

**Comment by pitrou** [comment]
`arrow-acero-util-test` hangs here:
```
#0  __futex_abstimed_wait_common64 (private=0, cancel=true, abstime=0x0, op=393, expected=0, futex_word=0x555555621c20) at ./nptl/futex-internal.c:57
#1  __futex_abstimed_wait_common (cancel=true, private=0, abstime=0x0, clockid=0, expected=0, futex_word=0x555555621c20) at ./nptl/futex-internal.c:87
#2  __GI___futex_abstimed_wait_cancelable64 (futex_word=futex_word@entry=0x555555621c20, expected=expected@entry=0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=private@entry=0) at ./nptl/futex-internal.c:139
#3  0x00007ffff4365a41 in __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x555555621b70, cond=0x555555621bf8) at ./nptl/pthread_cond_wait.c:503
#4  ___pthread_cond_wait (cond=0x555555621bf8, mutex=0x555555621b70) at ./nptl/pthread_cond_wait.c:627
#5  0x00007ffff7421774 in std::condition_variable::wait<arrow::internal::ThreadPool::WaitForIdle()::<lambda()> >(std::unique_lock<std::mutex> &, struct {...}) (this=0x555555621bf8, __lock=..., __p=...) at /home/antoine/mambaforge/envs/pyarrow/x86_64-conda-linux-gnu/include/c++/12.3.0/condition_variable:102
#6  0x00007ffff741f612 in arrow::internal::ThreadPool::WaitForIdle (this=0x55555561e0d0) at /home/antoine/arrow/dev/cpp/src/arrow/util/thread_pool.cc:497
#7  0x00005555555cad6a in arrow::acero::MakePrimedThreadPool (num_threads=11) at /home/antoine/arrow/dev/cpp/src/arrow/acero/task_util_test.cc:67
#8  0x00005555555ccf33 in arrow::acero::TaskScheduler_StressTwo_Test::TestBody (this=0x55555561a930) at /home/antoine/arrow/dev/cpp/src/arrow/acero/task_util_test.cc:193
#9  0x00007ffff7d686d1 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (object=0x55555561a930, method=&virtual testing::Test::TestBody(), location=0x7ffff7d26f83 "the test body") at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:2607
#10 0x00007ffff7d5f691 in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=0x55555561a930, method=&virtual testing::Test::TestBody(), location=0x7ffff7d26f83 "the test body") at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:2643
#11 0x00007ffff7d39b6f in testing::Test::Run (this=0x55555561a930) at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:2682
#12 0x00007ffff7d3a625 in testing::TestInfo::Run (this=0x55555561c2f0) at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:2861
#13 0x00007ffff7d3b011 in testing::TestSuite::Run (this=0x55555560b180) at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:3015
#14 0x00007ffff7d4b7d9 in testing::internal::UnitTestImpl::RunAllTests (this=0x55555561aa50) at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:5855
#15 0x00007ffff7d69690 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x55555561aa50, method=(bool (testing::internal::UnitTestImpl::*)(class testing::internal::UnitTestImpl * const)) 0x7ffff7d4b3c2 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x7ffff7d279e8 "auxiliary test code (environments or event listeners)") at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:2607
#16 0x00007ffff7d60878 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x55555561aa50, method=(bool (testing::internal::UnitTestImpl::*)(class testing::internal::UnitTestImpl * const)) 0x7ffff7d4b3c2 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x7ffff7d279e8 "auxiliary test code (environments or event listeners)") at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:2643
#17 0x00007ffff7d49be3 in testing::UnitTest::Run (this=0x7ffff7d80ec0 <testing::UnitTest::GetInstance()::instance>) at /build/build-test/_deps/googletest-src/googletest/src/gtest.cc:5438
#18 0x00007ffff7d82a02 in RUN_ALL_TESTS () at /build/build-test/_deps/googletest-src/googletest/include/gtest/gtest.h:2490
#19 0x00007ffff7d82a4b in main (argc=1, argv=0x7fffffffc108) at /build/build-test/_deps/googletest-src/googletest/src/gtest_main.cc:52
```

**Comment by pitrou** [comment]
I think the bottom line is that, for now, we cannot enable mold by default.

**Comment by kou** [comment]
The warnings are caused by sharing `cpp/src/arrow/symbols.map` with `libarrow*.so`.
`descriptor_table_Flight_2eproto` exists only in `libarrow_flight.so` and `descriptor_table_FlightSql_2eproto` exists only in `libarrow_flight_sql.so`.
I used `*` to suppress the warnings.

**Comment by kou** [comment]
These test problems aren't happen on my environment. (`g++ (Debian 13.2.0-13) 13.2.0`)
But mold doesn't work with conda. I'll disable this by default.

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp -g linux

**Comment by kou** [comment]
Hmm. It seems that mold doesn't work with sanitizers. I disabled `ARROW_USE_MOLD` if one of `ARROW_USE_{ASAN,TSAN,UBSAN}` is specified.

File: cpp/cmake_modules/DefineOptions.cmake [hunk 1]
```diff
@@ -110,6 +110,12 @@ macro(resolve_option_dependencies)
   if(MSVC_TOOLCHAIN)
     set(ARROW_USE_GLOG OFF)
   endif()
+  # Tests are crashed with mold + sanitizer checks.
+  if(ARROW_USE_ASAN
+     OR ARROW_USE_TSAN
+     OR ARROW_USE_UBSAN)
+    set(ARROW_USE_MOLD OFF)
```
Comment (assignUser) [inline cpp/cmake_modules/DefineOptions.cmake:117]: Should this not produce a message in case the user switched it on?

Comment (kou) [reply cpp/cmake_modules/DefineOptions.cmake]: OK. I've added a warning message for this case.

**Comment by kou** [comment]
I'll merge this tomorrow if nobody objects this.

**Review by assignUser — APPROVED** [review]
Cool!

**Pull request merged**